### ISSUE TITLE
Bazel: provide pre-built binary toolchain for protoc

### ISSUE THREAD

**Issue by alexeagle** [protocolbuffers/protobuf#19558]
From https://protobuf.dev/news/2024-10-01/#end-goal:

> Once the rules are in the protobuf repo, we intend to address common user requests, such as using prebuilts for the proto compiler where possible.

This is that request.

**What language does this apply to?**
All

**Describe the problem you are trying to solve.**

All Bazel users are expected to build `protoc` from source as a `cc_binary`. This leads to problems which are often reported on the Bazel Slack:

1. Bazel doesn't include a hermetic C++ toolchain, so the compilation fails for a subset of developers due to the host toolchain on their computer. This can be easily reproduced by registering a [non-functional toolchain](https://github.com/aspect-build/toolchains_protoc/blob/main/examples/tools/toolchains/BUILD.bazel). Many users have no C++ code, so they have no benefit from dealing with this hermeticity failure other than to repair `protoc`.
2. `protoc` frequently gets recompiled rather than being a cache hit - [example report](https://fzakaria.com/2024/10/23/bazel-knowledge-mind-your-path.html), [issue](https://github.com/bazelbuild/bazel/issues/7095). This makes Bazel builds slow.

**Describe the solution you'd like**

Bazel's toolchain feature allows it to download the pre-built binaries from [protobuf releases](https://github.com/protocolbuffers/protobuf/releases)

https://github.com/bazelbuild/rules_proto/pull/205 was part of my earlier work to provide this capability. The ruleset mirrors its own integrity hashes as part of each release.

https://github.com/aspect-build/toolchains_protoc/  is a user-land implementation of this proposal, however it was broken by changes in Bazel 8 and rules_proto described by https://protobuf.dev/news/2024-10-01/.

**Additional context**
Some user reports:
- https://fzakaria.com/2024/11/28/bazel-knowledge-protobuf-is-the-worst-when-it-should-be-the-best.html

I describe some GitHub Actions workflows for automating this pattern on https://blog.aspect.build/releasing-bazel-rulesets-rust (and the earlier https://blog.aspect.build/releasing-bazel-rulesets)

**Comment by shaod2** [comment]
Yes, we have changes pending to provide prebuilt protocs in the open source realm. I don't have the exact timeline tho; currently expecting for H1 2025

**Comment by vorburger** [comment]
Having to constantly rebuild `protoc` from scratch everywhere causes all sorts of weird surprises; e.g. https://github.com/grpc/grpc-java/issues/11790!

It also makes builds more "heavy" than they could otherwise be, which is a problem in certain more "light" (lite?) build environments; note e.g. https://github.com/jitpack/jitpack.io/issues/3129.

It would be really cool if using Protocol Buffers with Bazel would no longer require building `protoc`.

**Comment by mering** [comment]
In setups with RBE available, this is not a problem since it is mostly just downloaded from the remote cache anyways. In fact, in such setups it is usually preferred to build from source instead of downloading random binaries from the Internet for compliance reasons. So building from source should definitely be kept as an option and downloading pre built binaries should be optional.

**Comment by vorburger** [comment]
> In setups with RBE available, this is not a problem since it is mostly just downloaded from the remote cache anyways. 

Sure, but not every user of Protobuf of Bazel has RBE set-up for every project.

> In fact, in such setups it is usually preferred to build from source instead of downloading random binaries from the internet for compliance reasons.

This can of course easily be solved with some sort of checksum / hash that's verified on the download; à la [`http_archive's sha256`](https://bazel.build/rules/lib/repo/http#http_archive-sha256) or the [HTML's Subresource Integrity (SRI)](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) or that `?hl=` "standard" idea for [Cryptographic Hyperlinks from draft-sporny-hashlink-07](https://datatracker.ietf.org/doc/html/draft-sporny-hashlink-07) (what a shame that never gained more widespread traction) et al.

> So building from source should definitely be kept as an option and downloading pre built binaries should be optional.

An "opt in" flag (?) would already be a huge progress over the current situation!

**Comment by mering** [comment]
> This can of course easily be solved with some sort of checksum / hash that's verified on the download; à la [`http_archive's sha256`](https://bazel.build/rules/lib/repo/http#http_archive-sha256) or the [HTML's Subresource Integrity (SRI)](https://developer.mozilla.org/en-US/docs/Web/Security/Subresource_Integrity) or that `?hl=` "standard" idea for [Cryptographic Hyperlinks from draft-sporny-hashlink-07](https://datatracker.ietf.org/doc/html/draft-sporny-hashlink-07) (what a shame that never gained more widespread traction) et al.

This is only a small part of the story. How do you verify that the binary doesn't contain malicious code? You need reproducible builds first. Then you need to build from source and verify that the binary you download is actually the artifact built from the source you are expecting. You need to do this for every version as inspecting binary diffs is not handy. When you build from source, you only need to check the source diff which is much easier to review.
Dependency attacks are a thing. This is how companies get hacked.

> > So building from source should definitely be kept as an option and downloading pre built binaries should be optional.
> 
> An "opt in" flag (?) would already be a huge progress over the current situation!

Preferably as default with an opt-out flag. I see efficient building from source instead of downloading tons of binaries the main advantage of Bazel compared to other build systems.

**Comment by alexeagle** [comment]
@mering yes, supply chain security is an important consideration here, both under Bazel and any other build system.

I think you're pointing out https://github.com/protocolbuffers/protobuf/issues/16165 again - since `protoc` downloads aren't provided along with a checksum, users have to compute one themselves. Any Bazel rules that fetch should include a checksum - this is true whether they fetch sources and then compile them, or a binary that's compiled by someone else. https://github.com/aspect-build/toolchains_protoc/blob/main/protoc/private/versions.bzl#L54 for example.

Then yes, it would be nice to have a proof of provenance, via some attestation published on protoc releases. We are adding these right now to modules on the BCR. It will be trivial for tools like [`tar`](https://github.com/aspect-build/bsdtar-prebuilt) where we use GitHub Actions and they have a feature for this.

> setups with RBE available, this is not a problem

This isn't true for most Bazel users, since they ship Macs to their developers and the remote cache contains Linux binaries only.

**Comment by mering** [comment]
> @mering yes, supply chain security is an important consideration here, both under Bazel and any other build system.
> 
> I think you're pointing out #16165 again - since `protoc` downloads aren't provided along with a checksum, users have to compute one themselves. Any Bazel rules that fetch should include a checksum - this is true whether they fetch sources and then compile them, or a binary that's compiled by someone else. https://github.com/aspect-build/toolchains_protoc/blob/main/protoc/private/versions.bzl#L54 for example.
> 
> Then yes, it would be nice to have a proof of provenance, via some attestation published on protoc releases. We are adding these right now to modules on the BCR. It will be trivial for tools like [`tar`](https://github.com/aspect-build/bsdtar-prebuilt) where we use GitHub Actions and they have a feature for this.

This requires trusting whoever is specifying the checksums. If you don't want to (or are not allowed to) blindly trust someone providing correct checksums but reviewing the code yourself, this is much easier when you build source code instead of comparing binaries (which usually also requires building the code in the first place).

> > setups with RBE available, this is not a problem
> 
> This isn't true for most Bazel users, since they ship Macs to their developers and the remote cache contains Linux binaries only.

Why are they not just using Linux for development in the first place if this is what they test and ship with their CI/CD? It's usually a bad idea to test something different from what you ship...

**Comment by michaelschuett-tomtom** [comment]
> 
> Why are they not just using Linux for development in the first place if this is what they test and ship with their CI/CD? It's usually a bad idea to test something different from what you ship...

Most of the time before bazel is even introduced to the environment you have existing machines and at least for the last 10 years of my career that has been macs. I personally would like a linux machine the same as my deploy environment however it is just not the reality for most workplaces.

**Comment by mering** [comment]
> > Why are they not just using Linux for development in the first place if this is what they test and ship with their CI/CD? It's usually a bad idea to test something different from what you ship...
> 
> Most of the time before bazel is even introduced to the environment you have existing machines and at least for the last 10 years of my career that has been macs. I personally would like a linux machine the same as my deploy environment however it is just not the reality for most workplaces.

Interesting. In the various companies across different industries, I have seen more Linux machines than Macs in the past 15 years (mostly without using Bazel).

**Comment by jschaf** [comment]
Still active.

**Comment by jguamie** [comment]
This has been prioritized. We will be working with @alexeagle and Aspect.build to complete this by EOY 2025.

**Comment by mering** [comment]
@jguamie @alexeagle will there continue to be an option (flag/transition) to build protoc from source? We need to keep using the current setup.

**Comment by jguamie** [comment]
@mering yes, this won't impact existing setups.

**Comment by gfrankliu** [comment]
> This has been prioritized. We will be working with [@alexeagle](https://github.com/alexeagle) and Aspect.build to complete this by EOY 2025.

Thanks for prioritizing this!
Let's make sure we add the support for multiple protoc versions, like what [rules_python did](https://rules-python.readthedocs.io/en/latest/toolchains.html#pinning-to-a-python-version). Here is the [use case](https://github.com/aspect-build/toolchains_protoc/issues/76) I discussed with @alexeagle

**Comment by aaliddell** [comment]
.

**Comment by jayaprabhakar** [comment]
Any progress on this? It would be nicer if bazel could just execute installed protoc version. Since bazel is not `hermetic`, as it depends on the installed C++ tool chain, why not depend on installed protoc directly? 
Are is it possible to create a genrule that can use the installed protoc version? For those that want hermatic builds, they should anyway use external build service or a dedicated host for their builds anyway.

**Comment by alexeagle** [comment]
Yes there is progress, the most recent protobuf release includes
Feat: update bazel central registry publish workflow (https://github.com/protocolbuffers/protobuf/commit/d5217fd016809b3436c30a780216ebd34f4bed2d)

which adds a `prebuilt_tool_integrity.bzl` file inside the distribution. Next I need to change the toolchain registrations to use it.

**Comment by gfrankliu** [comment]
@jguamie wrote:
> This has been prioritized. We will be working with [@alexeagle](https://github.com/alexeagle) and Aspect.build to complete this by EOY 2025.

EOY is fast approaching :) 
Can we still expect it this year? I am looking forward to using different protoc versions for different languages in our monorepo.

**Comment by ouillie** [comment]
Just managed to build a `proto_library` with the pre-built binary using the new flag. Thanks!

**Comment by jguamie** [comment]
@gfrankliu wrote:
> Can we still expect it this year? I am looking forward to using different protoc versions for different languages in our monorepo.

Yes! We are coordinating with the Bazel team now to get this launched with the Bazel 9 LTS release.

**Comment by FrankPortman** [comment]
> Just managed to build a `proto_library` with the pre-built binary using the new flag. Thanks!

@ouillie do you have a link to instructions on how to test this?

**Comment by ouillie** [comment]
Since there's no versioned release yet, I just used a `git_override` in my `MODULE.bazel` to pull in protobuf at the commit where the flag was introduced: See [this commit](https://github.com/vimana-cloud/vimana/commit/5e1e3ab76b28d9e5002fae84e7dd007aee07283e). The only files in that commit that would be relevant to others in general are `.bazelrc`, `MODULE.bazel`, `protobuf-bzl-visibility.patch` and `protobuf-linkarr-attr.patch` (I found I had to introduce a couple small patches to make it work).

**Comment by zhangskz** [comment]
Note this was released in v33.4 via the --@protobuf//bazel/toolchains:prefer_prebuilt_protoc flag

**Comment by agluszak** [comment]
there's a PR to flip the flag by default https://github.com/protocolbuffers/protobuf/pull/25313

AFAIK Bazel 9 is going to be released with protobuf 33.4 (that is, before the flag is flipped), so unfortunately the flag is still going to be needed

**Comment by jp-fizzbee** [comment]
Do you have the instruction on how to use it? With bazel 8.x and 9.x.

I have the `proto_library`, `go_proto_library` build targets. What should I change to depend on the pre built protoc?

**Comment by agluszak** [comment]
> Do you have the instruction on how to use it? With bazel 8.x and 9.x.
> 
> I have the `proto_library`, `go_proto_library` build targets. What should I change to depend on the pre built protoc?

1. make sure you load definitions from protobuf (not rules_proto) and that protobuf version is 33.4
2. add to .bazelrc
```
common --incompatible_enable_proto_toolchain_resolution
common --@com_google_protobuf//bazel/toolchains:prefer_prebuilt_protoc=true
```
and optionally
```
# Ensure that we don't accidentally build protobuf or gRPC
common --per_file_copt=external/.*protobuf.*@--PROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
common --host_per_file_copt=external/.*protobuf.*@--PROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
common --per_file_copt=external/.*grpc.*@--GRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
common --host_per_file_copt=external/.*grpc.*@--GRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
```

That's it :)

**Comment by loeffel-io** [comment]
> > Do you have the instruction on how to use it? With bazel 8.x and 9.x.
> > I have the `proto_library`, `go_proto_library` build targets. What should I change to depend on the pre built protoc?
> 
> 1. make sure you load definitions from protobuf (not rules_proto) and that protobuf version is 33.4
> 2. add to .bazelrc
> 
> ```
> common --incompatible_enable_proto_toolchain_resolution
> common --@com_google_protobuf//bazel/toolchains:prefer_prebuilt_protoc=true
> ```
> 
> and optionally
> 
> ```
> # Ensure that we don't accidentally build protobuf or gRPC
> common --per_file_copt=external/.*protobuf.*@--PROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
> common --host_per_file_copt=external/.*protobuf.*@--PROTOBUF_WAS_NOT_SUPPOSED_TO_BE_BUILT
> common --per_file_copt=external/.*grpc.*@--GRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
> common --host_per_file_copt=external/.*grpc.*@--GRPC_WAS_NOT_SUPPOSED_TO_BE_BUILT
> ```
> 
> That's it :)

amazing! any first informations on how much this changes the build time for an go proto grpc build?

edit: did some simple quick run

`bazel build //...  0.04s user 0.04s system 0% cpu 55.297 total` vs `bazel build //...  0.06s user 0.07s system 0% cpu 1:55.28 total` with `bazel clean //...` before each run

**Comment by mmohsenz** [comment]
I'm using py_proto_library, and it seems like it's still building the protoc binary instead of using pre-built ones. Is py_proto_library not expected to use the prebuilt binary?

I see this dependency chain to `@protobuf//:protoc`:
```
  "@protobuf//python:python_toolchain (680dcaf)"
  "@protobuf//python:python_toolchain (680dcaf)" -> "@protobuf//python:protobuf_python (680dcaf)"
  "@protobuf//python:protobuf_python (680dcaf)"
  "@protobuf//python:protobuf_python (680dcaf)" -> "@protobuf//python:well_known_types_py_pb2 (680dcaf)"
  "@protobuf//python:well_known_types_py_pb2 (680dcaf)"
  "@protobuf//python:well_known_types_py_pb2 (680dcaf)" -> "@protobuf//python:well_known_types_py_pb2_genproto (680dcaf)"
  "@protobuf//python:well_known_types_py_pb2_genproto (680dcaf)"
  "@protobuf//python:well_known_types_py_pb2_genproto (680dcaf)" -> "@protobuf//:protoc (14f9f01)"
  "@protobuf//:protoc (14f9f01)"
```

**Comment by alexeagle** [comment]
Already reported here: https://github.com/protocolbuffers/protobuf/issues/25453

The test for this https://github.com/protocolbuffers/protobuf/blob/fe27c0fd348639e4d1eb980dc9d750da53dc83c0/examples/example_without_cc_toolchain/BUILD.bazel didn't include dependencies on any well-known-types.
